src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,141 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final String INDEX_OUT_OF_BOUNDS = "Index is out of bounds!";
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public boolean add(T value) {
-        return false;
+        if (size == 0) {
+            tail = new Node<>(null, value, null);
+            head = tail;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
+        return true;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            Node<T> newNode = getNode(index);
+            Node<T> addedValue = new Node<>(newNode.prev, value, newNode);
+            newNode.prev.next = addedValue;
+            newNode.prev = addedValue;
+        }
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T element : list) {
+            add(element);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> newNode = getNode(index);
+        T currentValue = newNode.item;
+        newNode.item = value;
+        return currentValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentValue = getNode(index);
+        final T currentElement = currentValue.item;
+        if (currentValue.next == null) {
+            tail = currentValue.prev;
+        } else {
+            currentValue.next.prev = currentValue.prev;
+        }
+        if (currentValue.prev == null) {
+            head = currentValue.next;
+        } else {
+            currentValue.prev.next = currentValue.next;
+        }
+        size--;
+        return currentElement;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentValue = head;
+        for (int i = 0; i < size; i++) {
+            if (object == currentValue.item || object != null && object.equals(currentValue.item)) {
+                remove(i);
```

Comment:
the reason is that in `remove()` method you go into another loop because it contains `getNode(index)` call
that loop is redundant here
```suggestion
                unlink(currentValue);
```